server/lib/oauth2/oauth2.executeOauth2HTTPQuery.js

```diff
@@ -0,0 +1,140 @@
+const queryString = require('query-string');
+const { ClientCredentials } = require('simple-oauth2');
+const { default: axios } = require('axios');
+const logger = require('../../utils/logger');
+const { OAUTH2 } = require('../../utils/constants');
+
+/**
+ * @description Refresh Oauth2 token in case of it expired .
+ * @param {Object} variable - Gladys variable manager.
+ * @param {string} serviceId - Gladys service id call method.
+ * @param {string} userId - Gladys userId of current session.
+ * @returns {Object} Response of oauth2 refresh token query .
+ * @example
+ * oauth2.refreshTokenAccess('7fdsf4s68r4gfr68f4r63csd7f6f4c3r85',
+ *    {...}, '78v4f3df83g74v1fsd8375f63gvrf5c');
+ */
+async function refreshTokenAccess(variable, serviceId, userId) {
+  const clientId = await variable.getValue(OAUTH2.VARIABLE.CLIENT_ID, serviceId, userId);
+  const secret = await variable.getValue(OAUTH2.VARIABLE.CLIENT_SECRET, serviceId, userId);
+  const accessToken = JSON.parse(await variable.getValue(OAUTH2.VARIABLE.ACCESS_TOKEN, serviceId, userId));
+
+  logger.trace('accessToken: ', accessToken);
+  // Find provider configuration
+  const tokenHost = await variable.getValue(OAUTH2.VARIABLE.TOKEN_HOST, serviceId);
+  const tokenPath = await variable.getValue(OAUTH2.VARIABLE.TOKEN_PATH, serviceId);
+  const authorizeHost = await variable.getValue(OAUTH2.VARIABLE.AUTHORIZE_HOST, serviceId);
+  const authorizePath = await variable.getValue(OAUTH2.VARIABLE.AUTHORIZE_PATH, serviceId);
+  const additionalAccessTokenRequestActionParam = await variable.getValue(
+    OAUTH2.VARIABLE.ADDITIONAL_ACCESS_TOKEN_REQUEST_ACTION_PARAM,
+    serviceId,
+  );
+
+  // Init credentials based on integration name
+  const credentials = {
+    client: {
+      id: clientId,
+      secret,
+    },
+    auth: {
+      tokenHost,
+      tokenPath,
+      authorizeHost,
+      authorizePath,
+    },
+    http: {
+      json: true,
+    },
+  };
+
+  try {
+    const client = new ClientCredentials(credentials);
+    let authResult = await client.createToken(accessToken);
+
+    if (authResult.expired()) {
```

Comment:
In this part I add 
this: 
```
 if (additionalAccessTokenRequestActionParam) {
          refreshParams.action = additionalAccessTokenRequestActionParam;
        }
```
and this : 
`if (!authResult.token.status || authResult.token.status === 0) {`

recently, only because withings WS change, at the begining I don't need those action (so this function is just to refresh token if expire)
I find weird to be add those specific case so the else case treat old method of withings WS because I thonk it's the common case .

Do you know if oauth2 refresh token need te specify action in common case ?
Do you know if the response contain always status and body ?

(I use simple-oauth2 for all oauth2 request)

Reply:
in simple-oauth2 lib it's necessary to do the code for refresh token 
https://www.npmjs.com/package/simple-oauth2#refresh-an-access-token